libcudacxx/test/libcudacxx/cuda/barrier/cp_async_bulk_ptx_compiles.pass.cpp

```diff
@@ -47,8 +47,8 @@ __global__ void test_bulk_tensor(CUtensorMap* map)
 __global__ void test_bulk(void* gmem)
 {
   __shared__ int smem;
-  __shared__ barrier bar;
-  cde::cp_async_bulk_global_to_shared(&smem, gmem, 1024, bar);
+  __shared__ barrier* bar;
+  cde::cp_async_bulk_global_to_shared(&smem, gmem, 1024, *bar);
```

Comment:
This looks highly suspicious: https://godbolt.org/z/4Ma1Yq9M4

We end up using an unintialized value passed to `cp.async.bulk.shared::cluster.global.mbarrier::complete_tx::bytes` via `%r3`

```
.visible .entry _Z9test_bulkPv(
        .param .u64 _Z9test_bulkPv_param_0
)
{
        .reg .b32       %r<6>;
        .reg .b64       %rd<4>;
        .loc    1 6 0
        .shared .align 4 .u32 _ZZ9test_bulkPvE4smem;

        ld.param.u64    %rd3, [_Z9test_bulkPv_param_0];
        cvta.to.global.u64      %rd1, %rd3;
        mov.u32         %r1, _ZZ9test_bulkPvE4smem;
        mov.u32         %r5, 1024;
        cp.async.bulk.shared::cluster.global.mbarrier::complete_tx::bytes [%r1], [%rd1], %r5, [%r3]; // 1a. unicast
```

Reply:
@miscco This is still an issue. We do need `bar` pointing to valid shared memory location. 

Using something like this appears to make the cp_async_bulk* tests work.
```
  alignas(128) __shared__ int smem_buffer[buf_len];
  __shared__ char barrier_data[sizeof(barrier)];
  barrier &bar = *reinterpret_cast<barrier*>(&barrier_data);
```

Reply:
I guess this may be considered to be a new bug. I'll file the issue and send the patch tomorrow. LMK if this workaround is OK. I may still need to look at why clang complains about non-trivial initializer for `barrier`.

Reply:
I believe this is an old version, on main it lreads:
```
  __shared__ barrier* bar;
  if (threadIdx.x == 0)
  {
    init(bar, blockDim.x);
  }

  cde::cp_async_bulk_global_to_shared(&smem, gmem, 1024, *bar);
```

Reply:
The bar is still an uninitialized pointer and it's expected to be pointing to the valid data to initialized, and then used.
E.g. https://github.com/NVIDIA/cccl/blob/a92db6b1915d946b48a5e48013d5ca7fc898a67d/libcudacxx/include/cuda/__barrier/barrier_block_scope.h#L107

In the original comment I've quoted PTX the current code produces and it's clearly wrong. Re're passing `[%r3]` as the last `mbar` argument, but `r3` is not set to anything.

Reply:
Oh we are missing the sync I believe, because `init(bar, blockDim.x)` does initialize it but we need to ensure that we wait for that

Reply:
opened #3061

Reply:
> init(bar, blockDim.x) does initialize it

I'm confused (or missing something). 
`bar` is a pointer, pointing who knows where, right? If we pass an invalid pointer to init, what exactly does `init` initialize then? AFAICT init() assumes that `bar` points to valid storage in shared memory.

In the original code `bar` was a variable and that looks like what the barrier APIs expect:
https://github.com/NVIDIA/cccl/blob/baee3f502d4fd5febc7f602428592b2bec5d2d3b/libcudacxx/test/libcudacxx/cuda/barrier/cp_async_bulk_tensor.pass.cpp#L71-L75